### DISCUSSION TITLE
How to use the mat file generated by ants to register the TRK file to T1

### DISCUSSION THREAD

**Discussion by wangyibin0011** [dipy/dipy#2165] [Registration] [answered]
Hello, I have used antspy to generate the registration matrix file from DTI to T1, the file format is mat, at the same time, I generated the streamlines the same location of DTI, the file format is TRK, I want to know what method can I use the mat file to move the TRK file to the location of T1? I noticed that the coordinate system of T1 is LFS and the coordinate system of TRK is RAS. Thank you!

**Comment by arokem** [comment]
Hi! Thanks for asking here. What's in the mat file? Could you share one of these with us?

**Comment by wangyibin0011** [comment]
[T1_MNI.zip](https://github.com/dipy/dipy/files/4596679/T1_MNI.zip)
I opened the mat file in the following way:
import SimpleITK as sitk
Transform = sitk.ReadTransform(r"C:\Users\wyb\Desktop\need_file\T1_MNI.mat")
print(Transform)
![1](https://user-images.githubusercontent.com/53501301/81363667-79621300-9116-11ea-8f6a-22b3aa76aa68.png)

**Comment by arokem** [comment]
Hey - have you tried using the (transform_tracking_output)[https://github.com/dipy/dipy/blob/553d92ebdeda866d635ec414e767f22d9e252046/dipy/tracking/utils.py#L846] function?

You can get the `affine` input like so:

```
from scipy.io import loadmat
import numpy as np

mat = loadmat('T1_MNI.mat')
aff = np.eye(4)
aff[:3, :3] = mat['AffineTransform_float_3_3'][:9].reshape((3,3))
aff[:3, 3] =  mat['AffineTransform_float_3_3'][9:].squeeze()
```

**Comment by frheault** [comment, marked as answer]
Here is some of my personal code for this problem:
The matrix from ants isn't actually an affine like dipy, it is an itkMatrixOffsetTransformBase in LPS convention.
This code gives the same results as ants' `ConvertTransformFile 3 MATRIX.mat MATRIX.txt --hm --ras`

I used it with streamlines (`from dipy.tracking.streamline import transform_streamlines`), however you will have to inverse the matrix as Ants save its transformation inverted so it is usable for image registration. Streamlines being 3D points the application isn't done the same way (just use `np.linalg.inv`)

```
def load_matrix_in_any_format(filepath):
    _, ext = os.path.splitext(filepath)
    if ext == '.txt':
        data = np.loadtxt(filepath)
    elif ext == '.npy':
        data = np.load(filepath)
    elif ext == '.mat':
        # .mat are actually dictionnary. This function support .mat from
        # antsRegistration that encode a 4x4 transformation matrix.
        transfo_dict = loadmat(filepath)
        lps2ras = np.diag([-1, -1, 1])

        rot = transfo_dict['AffineTransform_double_3_3'][0:9].reshape((3, 3))
        trans = transfo_dict['AffineTransform_double_3_3'][9:12]
        offset = transfo_dict['fixed']
        r_trans = (np.dot(rot, offset) - offset - trans).T * [1, 1, -1]

        data = np.eye(4)
        data[0:3, 3] = r_trans
        data[:3, :3] = np.dot(np.dot(lps2ras, rot), lps2ras)
    else:
        raise ValueError('Extension {} is not supported'.format(ext))

    return data
```
